File: day02/Day02.scala
import scala.io.Source.stdin

@main def main(idx: Int) =
  val xs = stdin.getLines().map{_.split(" ")}.map{x => (x(0), x(1).toInt)}

  if idx == 1 then
    var x = 0; var y = 0
    for ((dir, n) <- xs)
      dir match {
        case "forward" => x += n
        case "down" => y += n
        case "up" => y -= n
      }
    println(x * y)
  else
    var x = 0; var y = 0; var aim = 0
    for ((dir, n) <- xs)
      dir match {
        case "down" => aim += n
        case "up" => aim -= n
        case "forward" =>
          x += n
          y += n * aim
      }
    println(x * y)

File: day02/Day02.rs
#![feature(stdin_forwarders)]
use std::env;
use std::io;

fn main() {
    let stmts = io::stdin()
        .lines()
        .map(|s| {
            let s = s.ok()?;
            let mut w = s.split(" ");

            let dir = w.next()?.to_owned();
            let n = w.next()?.parse::<u32>().ok()?;

            Some((dir, n))
        })
        .map(Option::unwrap);

    let (mut x, mut y, mut aim) = (0, 0, 0);

    if env::args().nth(1).unwrap() == "1" {
        for inst in stmts {
            match (inst.0.as_str(), inst.1) {
                ("forward", n) => x += n,
                ("down", n) => y += n,
                ("up", n) => y -= n,
                _ => (),
            }
        }
    } else {
        // Part two
        for inst in stmts {
            match (inst.0.as_str(), inst.1) {
                ("down", n) => aim += n,
                ("up", n) => aim -= n,
                ("forward", n) => {
                    x += n;
                    y += aim * n;
                }
                _ => (),
            }
        }
    }

    println!("{}", x * y);
}
